server/routerlicious/packages/routerlicious-base/src/riddler/tenantManager.ts

```diff
@@ -236,6 +247,16 @@ export class TenantManager {
      * Retrieves the secret for the given tenant
      */
     public async getTenantKeys(tenantId: string, includeDisabledTenant = false): Promise<ITenantKeys> {
+        const lumberProperties = { [BaseTelemetryProperties.tenantId]: tenantId };
+
+        // Read from cache first
+        const cachedKey = await this.cache.get(tenantId);
+        if (cachedKey) {
+            Lumberjack.info(`Tenant key(s) found in cache`, lumberProperties);
+            return this.decryptCachedKeys(cachedKey);
+        }
+
+        // Read from database if keys aren't found in the cache
         const tenantDocument = await this.getTenantDocument(tenantId, includeDisabledTenant);
```

Comment:
Since tenant keys are a property of the tenantDocument, and there are many more uses of retrieving tenantDocument than just keys, it seems like it would be more efficient to cache the entire tenantDocument rather than just the keys. That would also simplify the following key1/key2 retrieval validation code (none of the new additions would be necessary, since we are just using a tenantDocument like before this change). The changes instead would be within `getTenantDocument` (for read) and the related update functions `deleteTenant`, `refreshTenantKey`, `updateStorage`, `updateOrderer`, and `updateCustomData` (for cache invalidation).

This would also resolve the current (possible) bug in this PR that might allow a token for a deleted tenant to be validated, because we rely on `getTenantDocument` returning null to know if a tenant is disabled or deleted.

Reply:
@yangg-msft What are your thoughts about storing the entire tenantDocument rather than just the tenantKeys? The initial plans were to store the keys only, but would there be more benefits to storing the tenantDocument?

Reply:
Not sure I understand your point @zhangxin511. Maybe this is a naive question, but what role is the red dots (storage units in our service architecture) playing here? What I think @znewton referred to was the [ITenantDocument](https://github.com/microsoft/FluidFramework/blob/9d29441a8056c60bffe4ada18e700617fc085497/server/routerlicious/packages/routerlicious-base/src/riddler/tenantManager.ts#L26) type, which has metadata about a document and does not include the document content we put in storage. Instead, my understanding was that we kept `ITenantDocument` data in the the Mongo/CosmosDB only... And I guess in this case, that would be the group DB. Is that not the case?

Reply:
@hedasilv , @znewton , @kekachmar Sorry for my english and my terminology. The reason I mention red dot is I have the impression that single resource is the red dot, for example, the real red dot and where I have the bad conception for group cosmos DB.

But other that, there is no way for the blue dots knows ITenantDocument got updated beside the one who handled the resource by our current design, only the tenant key we can inferred from 403. That is what this PR is about.

Reply:
I think to stay within the scope of this work item, we can cache the tenantKeys only for now